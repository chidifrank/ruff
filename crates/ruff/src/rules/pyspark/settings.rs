//! Settings for the `pyspark` plugin.

use serde::{Deserialize, Serialize};

use ruff_macros::{CacheKey, CombineOptions, ConfigurationOptions};

#[derive(
    Debug, PartialEq, Eq, Serialize, Deserialize, Default, ConfigurationOptions, CombineOptions,
)]
#[serde(
    deny_unknown_fields,
    rename_all = "kebab-case",
    rename = "PySparkOptions"
)]
#[cfg_attr(feature = "schemars", derive(schemars::JsonSchema))]
pub struct Options {
    #[option(
        default = "3",
        value_type = "int",
        example = r#"
            # Flag errors (`PS001.py`) whenever the complexity level exceeds 3.
            max-complexity = 3
        "#
    )]
    /// The maximum PySpark transformation complexity to allow before triggering `PS001.py` errors.
    pub max_complexity: Option<usize>,
}

#[derive(Debug, CacheKey)]
pub struct Settings {
    pub max_complexity: usize,
}

const DEFAULT_MAX_COMPLEXITY: usize = 3;

impl Default for Settings {
    fn default() -> Self {
        Self {
            max_complexity: DEFAULT_MAX_COMPLEXITY,
        }
    }
}

impl From<Options> for Settings {
    fn from(options: Options) -> Self {
        Self {
            max_complexity: options.max_complexity.unwrap_or(DEFAULT_MAX_COMPLEXITY),
        }
    }
}

impl From<Settings> for Options {
    fn from(settings: Settings) -> Self {
        Self {
            max_complexity: Some(settings.max_complexity),
        }
    }
}
