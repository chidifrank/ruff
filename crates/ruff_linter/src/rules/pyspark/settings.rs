//! Settings for the `pyspark` plugin.

use crate::display_settings;
use ruff_macros::CacheKey;
use std::fmt::{Display, Formatter};

#[derive(Debug, Clone, CacheKey)]
pub struct Settings {
    pub max_complexity: usize,
}

pub const DEFAULT_MAX_COMPLEXITY: usize = 3;

impl Default for Settings {
    fn default() -> Self {
        Self {
            max_complexity: DEFAULT_MAX_COMPLEXITY,
        }
    }
}

impl Display for Settings {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        display_settings! {
            formatter = f,
            namespace = "linter.pyspark",
            fields = [
                self.max_complexity
            ]
        }
        Ok(())
    }
}
