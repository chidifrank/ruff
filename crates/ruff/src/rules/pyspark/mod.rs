pub(crate) mod rules;
pub mod settings;

#[cfg(test)]
mod tests {
    use std::path::Path;

    use anyhow::Result;
    use test_case::test_case;

    use crate::assert_messages;
    use crate::registry::Rule;
    use crate::settings::Settings;
    use crate::test::test_path;

    #[test_case(0)]
    #[test_case(3)]
    #[test_case(5)]
    fn max_complexity_spark(max_complexity: usize) -> Result<()> {
        let snapshot = format!("max_complexity_{max_complexity}");

        let diagnostics = test_path(
            Path::new("pyspark/PS001.py"),
            &Settings {
                pyspark: super::settings::Settings { max_complexity },
                ..Settings::for_rules(vec![Rule::SparkComplexStructure])
            },
        )?;

        assert_messages!(snapshot, diagnostics);
        Ok(())
    }
}
