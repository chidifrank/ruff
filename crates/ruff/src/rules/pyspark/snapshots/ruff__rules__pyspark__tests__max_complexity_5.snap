---
source: crates/ruff/src/rules/pyspark/mod.rs
---

